### PR TITLE
[fix] FSDP: fix the corner case of all params are in the children

### LINKED ISSUE #437: [FSDP] breaks when all modules within FSDP root are wrapped
## 🐛 Bug

When all child modules that have trainable parameters are wrapped with FSDP, this breaks. I spent some time digging to figure out the issue till I came to this

## To Reproduce

To reproduce in the test file here, you have to change a line and add a new test:

Wrap all layers instead of just some:

https://github.com/facebookresearch/fairscale/blob/master/tests/nn/data_parallel/test_fsdp.py#L691

```python
class NestedWrappedModule(nn.Module):
    def __init__(self, group, wrapper_config):
        super().__init__()
        self.rank = group.rank()
        self.world_size = group.size()
        self.wrapper_config = wrapper_config

        def _maybe_wrap(layer):
            if wrapper_config is not None:
                return FullyShardedDataParallel(layer, group, **wrapper_config)
            return layer

        torch.manual_seed(0)  # keep everything deterministic
        self.module = nn.Sequential(
            _maybe_wrap(nn.Linear(8, 4)), _maybe_wrap(nn.Linear(4, 16)), _maybe_wrap(nn.Linear(16, 4)), _maybe_wrap(nn.Linear(4, 8)),
        )
```

I made a test new like below to run:

```python
class TestComparisonToPyTorchDDP(DistributedTest):
    ...
    def test_nested_all_wrapped_model(self):
        # We use a model with a long CUDA delay right before the optimizer step.
        # This tests our streams logic, and that we don't start the FP32 -> FP16
        # transfer until after the optimization step completes.
        config = {"mixed_precision": True}
        model_fn = NestedWrappedModule
        test_fn = functools.partial(self._test_identical_outputs, model_fn, config)
        spawn_and_init(test_fn)
```

## Expected behavior

I'm unsure if this a limitation of FSDP and it should work. It isn't the highest priority since I think in most cases you wouldn't wrap everything single trainable module in FSDP I think!

### PR DESCRIPTION
**Pull request by min-xu-ai**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
Fixes #437 .

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by min-xu-ai** [comment]
failure:

```
E
E       -- Process 0 terminated with the following error:
E       Traceback (most recent call last):
E         File "/home/owen/e/py39_clip/lib/python3.9/site-packages/torch/multiprocessing/spawn.py", line 59, in _wrap
E           fn(i, *args)
E         File "/home/owen/git/fairscale/tests/nn/data_parallel/test_fsdp.py", line 816, in init_and_run
E           fn(rank, group, *args)
E         File "/home/owen/git/fairscale/tests/nn/data_parallel/test_fsdp.py", line 293, in _test_identical_outputs
E           shard_loss = cls._train_for_several_steps(model, num_steps, autocast, lr=lr, norm_type=norm_type)
E         File "/home/owen/git/fairscale/tests/nn/data_parallel/test_fsdp.py", line 64, in _train_for_several_steps
E           model.clip_grad_norm_(clip_norm, norm_type)
E         File "/home/owen/e/py39_clip/lib/python3.9/site-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
E           return func(*args, **kwargs)
E         File "/home/owen/git/fairscale/fairscale/nn/data_parallel/fully_sharded_data_parallel.py", line 240, in clip_grad_norm_
E           assert self.assert_state(TrainingState.IDLE)
E         File "/home/owen/git/fairscale/fairscale/nn/data_parallel/fully_sharded_data_parallel.py", line 971, in assert_state
E           assert (
E       AssertionError: expected to be in state TrainingState.IDLE but current state is TrainingState.BACKWARD

../../e/py39_clip/lib/python3.9/site-packages/torch/multiprocessing/spawn.py:150: ProcessRaisedException
---------------------------------------------------------------------- Captured stdout call -----------------------------------------------------------------------
dist init r=0, world=1
XXX1
XXX1
XXX2
False False
XXX3
XXX1
XXX2
False False
XXX3
XXX1
XXX2
False False
XXX3
XXX1
XXX2
False False
XXX3
===================================================================== short test summary info =====================================================================
FAILED tests/nn/data_parallel/test_fsdp.py::TestComparisonToPyTorchDDP::test_nested_all_wrapped_model - torch.multiprocessing.spawn.ProcessRaisedException:
```

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -190,8 +190,9 @@ def __init__(
 
         # Flag to indicate if we require gradient reduction in the backward
         # pass. This will be False when inside the no_sync context manager.
-        self.require_backward_grad_sync: bool = True
+        self._require_backward_grad_sync: bool = True
```
Comment (min-xu-ai) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:193]: make this private, by the way.

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 2]
```diff
@@ -235,8 +236,10 @@ def clip_grad_norm_(
         .. warning:: This needs to be called on all ranks, since synchronization
             primitives will be used.
         """
+        torch.cuda.synchronize()
+        self._lazy_init()
```
Comment (min-xu-ai) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:240]: added these two lines to be safe

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 3]
```diff
@@ -94,6 +94,13 @@ class FullyShardedDataParallel(nn.Module):
         since FSDP will shard parameters in-place and this will break any
         previously initialized optimizers.
 
+    .. warning::
+
+        If you wrap every parameter inside a nested FSDP and leaving the outer
+        FSDP empty without any parameter, checkpointing activation may trigger
+        an assert on the backward pass. The solution is to leave some parameters
+        to the outer FSDP.
```
Comment (myleott) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:102]: oh nice! This was the issue you ran into previously right? I wonder if we can hack this with a zero-element param on the outer-most FSDP instance: `self.params = [Parameter(torch.empty(0))]` or something like that

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 4]
```diff
@@ -661,14 +672,21 @@ def _set_is_root(self) -> None:
...
+        self._queue_wait_for_post_backward_closure = self._queue_wait_for_post_backward
+        # As the root, we now set all children instances to False and
+        # give them a closure to try to queue a wait_for_post_backward.
         self.children_share_process_group = True
         for n, m in self.named_modules():
+            # `n != ""` excludes self.
             if n != "" and isinstance(m, FullyShardedDataParallel):
                 assert m._is_root is None
                 m._is_root = False
+                if not self._has_params:
```
Comment (myleott) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:687]: Add a comment?
```
# When root instance doesn't have params, allow children instances
# to queue the post_backward hook.
```

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 5]
```diff
@@ -779,7 +797,11 @@ def _register_post_backward_hooks(self) -> None:
         """Register backward hooks to reshard params and reduce-scatter grads."""
         if not torch.is_grad_enabled():
             return  # don't register grad hooks if grad isn't enabled
-        self._post_backward_callback_queued = False
+        if self._is_root:
+            # This actually means that only root instance has this field
+            # defined. Accidentally accessing this field will assert on all
+            # other instances, giving us a nice bug checker.
+            self._post_backward_callback_queued = False
```
Comment (myleott) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:804]: Nice! 😄

**Review by myleott — APPROVED** [review]

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py (hunk 3 above)
Comment (min-xu-ai) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:102]: I was thinking about it but I don't know how to hook it into the user's compute graph though. Perhaps keep that as a separate PR when we have an idea? I will add a todo.

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py (hunk 2 above)
Comment (min-xu-ai) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:263]: will do and will add a comment

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py (hunk 4 above)
Comment (min-xu-ai) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:687]: will do

**Pull request merged**